torchao/prototype/calibration_flow/static_quant.py

```diff
@@ -0,0 +1,107 @@
+        calibrating_linear.weight = float_linear.weight
+        calibrating_linear.bias = float_linear.bias
+        return calibrating_linear
+
+class QuantizedLinear(torch.nn.Module):
+    def __init__(self, in_features: int, out_features: int, act_obs: torch.nn.Module, weight_obs: torch.nn.Module, weight: torch.Tensor, bias: torch.Tensor):
+        super().__init__()
+        self.act_scale, self.act_zero_point = act_obs.calculate_qparams()
+        weight_scale, weight_zero_point = weight_obs.calculate_qparams()
+        assert weight.dim() == 2
+        block_size = (1, weight.shape[1])
+        target_dtype = torch.uint8
+        self.qweight = to_affine_quantized_static(weight, weight_scale, weight_zero_point, block_size, target_dtype)
+        self.bias = bias
+
+    def forward(self, input: Tensor):
+        block_size = input.shape
+        target_dtype = torch.uint8
+        qinput = to_affine_quantized_static(input, self.act_scale, self.act_zero_point, block_size, target_dtype)
+        return F.linear(qinput, self.qweight, self.bias)
+
+    @classmethod
+    def from_calibrating(cls, calibrating_linear):
+        quantized_linear = cls(calibrating_linear.in_features, calibrating_linear.out_features, calibrating_linear.act_obs, calibrating_linear.weight_obs, calibrating_linear.weight, calibrating_linear.bias)
+        return quantized_linear
+
+class ToyLinearModel(torch.nn.Module):
+    def __init__(self, m=64, n=32, k=64):
+        super().__init__()
+        self.linear1 = torch.nn.Linear(m, n, bias=False)
+        self.linear2 = torch.nn.Linear(n, k, bias=False)
+
+    def example_inputs(self, batch_size=1, dtype=torch.float32, device="cpu"):
+        return (torch.randn(batch_size, self.linear1.in_features, dtype=dtype, device=device),)
+
+    def forward(self, x):
+        x = self.linear1(x)
+        x = self.linear2(x)
+        return x
+
+dtype = torch.bfloat16
+m = ToyLinearModel(1024, 1024, 1024).eval().to(dtype).to("cuda")
+m_bf16 = copy.deepcopy(m)
+example_inputs = m.example_inputs(dtype=dtype, device="cuda")
+
+m_bf16 = torch.compile(m_bf16, mode='max-autotune')
+
+from torchao.quantization.quant_api import _replace_with_custom_fn_if_matches_filter
+import copy
+
+def to_calibrating_(model, act_obs, weight_obs):
+    _is_linear = lambda m, fqn: isinstance(m, torch.nn.Linear)
+    replacement_fn = lambda m: CalibratingLinear.from_float(m, act_obs, weight_obs)
+    act_obs = copy.deepcopy(act_obs)
+    weight_obs = copy.deepcopy(weight_obs)
+    _replace_with_custom_fn_if_matches_filter(model, replacement_fn, _is_linear)
+
+def to_quantized_(model):
+    _is_calibrating_linear = lambda m, fqn: isinstance(m, CalibratingLinear)
+
+    replacement_fn = lambda m: QuantizedLinear.from_calibrating(m)
+    _replace_with_custom_fn_if_matches_filter(model, replacement_fn, _is_calibrating_linear)
+
+act_obs = MinMaxObserver(dtype=torch.uint8, qscheme=torch.per_tensor_affine).to("cuda")
```

Comment:
i'd vote for rewriting this stuff without having to use old concepts like `qscheme`, instead of trying to reuse the code

Reply:
agree, this is just an example, the end goal is to implement a generic observer for blockwise quantization

Reply:
What's the blocker for getting rid of the torch.ao dependency now? Are there toy observers we can include for now?

Reply:
I think we want to define a general observer that works for the new quant primitives in the end so that we can replace this.

we can define a toy observer I think. although I'm not sure why we want to get rid of torch.ao dep since we already depend on pytorch.

Reply:
It's more that torchao is supposed to be a full replacement. Id make an exception if some piece of code is very large and complex but that doesn't seem to be the case here